### ISSUE TITLE
"Selective" Query Caching

### ISSUE THREAD

**Issue by timkelty** [markhuot/craftql#150]
I'm sad to say, I have a CraftQL in production in the wild that is has become apparent needs some level of query caching.

Here's what I'm thinking for the most basic-level of caching:

- Nothing is cached automatically
- No DB interaction at all - responses would be cached only with Craft/Yii cache component (`Craft::$app->getCache()`)
- The cache key would be provided as part of the request (perhaps using a `Surrogate-Key` header)?

I'll probably have to work on a fork with this soon, so any input is appreciated!

**Comment by timkelty** [comment]
Alternatively the cache key could be an md5 of the request body, but I almost like it better being totally manual.

**Comment by markhuot** [comment]
Hey Tim, can you provide any more detail about where things are getting particularly slow for you? I have a branch that switched GraphQL over to async fetching which does query batching and avoids the N+1 issues the default implementation handles. That might be enough for you, but I need to know what you're seeing slow down to provide more help.

**Comment by timkelty** [comment]
You're referring to https://github.com/markhuot/craftql/tree/batch-queries?
I'll give that branch a shot and compare.

I'll see if I can provide more details – my situation is a bit tricky as I'm merely providing an API that a separate agency is consuming…and I'm now getting reports of "slow responses".

What I'm pretty sure is going on, is they are executing pretty massive "page" queries, where they get the entry of a page, which has alot of matrix content, which has blocks with lots of related elements – so yes, an n+1 issue. If this were a twig template, eager loading would be the first thing I'd try.

**Comment by markhuot** [comment]
Yup, batching _should_ help here. If there's any way you could get me a sample query or some sort of test I could run myself I could definitely help debug this. If you can't post it here feel free to email it to me directly or reach out on the CraftCMS Slack.

**Comment by markhuot** [comment]
Sorry, @timkelty, one last note. Check out the https://github.com/markhuot/craftql/tree/batch-queries-defer branch. It's more up to date then the one you mention. The `-defer` branch has the N+1 solution.

**Comment by timkelty** [comment]
@markhuot so I'm running off that branch now, but not sure what I should be expecting…

When I peeked at the Api controller, it seems like requests were still `$singular: true`, so not batched…
Am I supposed to interact differently with the API to use batched requests?

**Comment by markhuot** [comment]
Oh, so I'm trying to make lives easier with that. Basically, if you pass a single query you're going to get a single query back. Typically if you pass a batch it has to be an array, like this,

```json
[
  {"query": "...", "variables": {}},
  {"query": "...", "variables": {}},
]
```

But I wrote it such that you can turn on batching and continue sending a single query without issue,

```json
{
  "query": "...",
  "variables": {},
}
```

Regardless of the initial query you send to the server you should see batching benefits from things like,

```graphql
{
  recentEntries: entriesConnection(limit:10) {
    edges {
      node {
        title
        author {
          name
        }
      }
      relatedEntries {
        edges {
          node {
            title
            author {
              name
            }
          }
        }
      }
    }
  }
  me: users(id: $id) {
    name
  }
}
```

The above is going to pull in recent entries, plus related entries, plus a user. Let's say you only have a few authors so the top level entry pulls back eight Tim users and the `relatedEntries` pulls back 4 Tim users and `me` is a Tim user. In that case the batching will make _one_ query for all the authors across all of the selection sets.

**Comment by timkelty** [comment]
ah gotcha! thanks. More testing coming!

**Comment by membla** [comment]
@timkelty Did you manage to make any significant improvements to response time, either through one of Marks other branches or a fork of your own? I too feel like I need to implement some sort of response caching... my heavy queries will have very static, cache-friendly responses so would love to implement some sort of server level caching for these. Not sure what would be the best approach though.

**Comment by timkelty** [comment]
@membla I did hack in some very basic response caching here: https://github.com/markhuot/craftql/compare/master...timkelty:dlr2018

I'm using it in production right now, purely out of necessity. It's super-hacky, so be forewarned. It is meant to be temporary.

It also listens for a `Surrogate-Key` header in the requests, and will tag the cache with that. That way you can listen on entry save, or whatever, and clear the cache for that item based on something from the entry (e.g. `uri`, `entryId`).

Feel free to hit me up on the Craft Slack if you need some more help.

**Comment by membla** [comment]
Tim's solution is functional in my setup too.
@markhuot Do you reckon some sort of response caching would make it to a release for the foreseeable future?
